unison-ucm: init at 1.0.M1c alpha

virusdave opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
Introducing the unison code manager.  Previously available via homebrew only.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

virusdave committed:
unison-ucm: init at 1.0.M1c alpha

virusdave commented:
@worldofpeace Any chance you can merge this?  Or point me to an appropriate person to ping?  Thanks!

worldofpeace commented on pkgs/development/compilers/unison/default.nix:
```diff
@@ -0,0 +1,43 @@
+{ stdenv, fetchurl, gnutar }:
+
+stdenv.mkDerivation rec {
+  name = "unison-code-manager-${version}";
```
```suggestion
  pname = "unison-code-manager";
```

worldofpeace commented on pkgs/development/compilers/unison/default.nix:
```diff
@@ -0,0 +1,43 @@
+      sha256 = "1iwynqnp1i39pyq9wc01x7y22y1qa0rrjlx40jjdgnj23y1r6jk4";
+    };
+
+  phases = "installPhase";
```
```suggestion
  dontBuild = true;
  dontConfigure = true;
```

worldofpeace replied:
we'd like to deprecate explicitly defining `phases`

worldofpeace commented on pkgs/development/compilers/unison/default.nix:
```diff
@@ -0,0 +1,43 @@
+
+  # There must be a simpler "just unpack and copy, dammit" standard derivation.
+  installPhase = ''
+    mkdir -p ${placeholder "out"}/bin
```
```suggestion
    mkdir -p $out/bin
```

I don't see this done typically in `installPhase`

worldofpeace commented on pkgs/development/compilers/unison/default.nix:
```diff
@@ -0,0 +1,43 @@
+  installPhase = ''
+    mkdir -p ${placeholder "out"}/bin
+
+    ${gnutar}/bin/tar -zxf $src -C ${placeholder "out"}/bin
```
Whatever got unpacked should be in the $CWD for you to copy into $out

worldofpeace replied:
```
cp ucm $out/bin
```

worldofpeace commented on pkgs/development/compilers/unison/default.nix:
```diff
@@ -0,0 +1,43 @@
+    maintainers = [
+      {
+        email = "dave.nicponski@gmail.com";
+        github = "virusdave";
```
you need to add yourself to `maintainers/maintainers-list.nix`

worldofpeace commented on pkgs/development/compilers/unison/default.nix:
```diff
@@ -0,0 +1,43 @@
+    description = "The Unison codebase manager & programming environment";
+    homepage = http://unisonweb.org/posts/ ;
+
+    license = stdenv.lib.licenses.bsd3;
```
```suggestion
    license = licenses.bsd3;
```

worldofpeace commented on pkgs/development/compilers/unison/default.nix:
```diff
@@ -0,0 +1,43 @@
+  meta = with stdenv.lib; {
+    description = "The Unison codebase manager & programming environment";
+    homepage = http://unisonweb.org/posts/ ;
+
```
no whitespace here

worldofpeace commented on pkgs/development/compilers/unison/default.nix:
```diff
@@ -0,0 +1,43 @@
+  '';
+
+  meta = with stdenv.lib; {
+    description = "The Unison codebase manager & programming environment";
```
Don't restate the software name in `description` just describe it.

worldofpeace replied:
Their github says
```
Next generation programming language, currently in development
```

why not this?

worldofpeace replied:
or maybe?
```
Modern, statically-typed purely functional language
```

also from github

worldofpeace commented on pkgs/development/compilers/unison/default.nix:
```diff
@@ -0,0 +1,43 @@
+      }
+    ];
+
+    platforms = platforms.linux ++ platforms.darwin;
```
```suggestion
    platforms = [ "x86_64-darwin" "x86_64-linux" ];
```

binary package that only supports these platforms

worldofpeace reviewed (changes requested):
This is a binary package so it needs to be patched.
You should be able to do this by adding `autoPatchelfHook` to `nativeBuildInputs`.

worldofpeace commented on pkgs/development/compilers/unison/default.nix:
```diff
@@ -0,0 +1,41 @@
+  # to not have "no new directories created" as a fatal condition.
+  dontBuild = true;
+  dontConfigure = true;
+  phases = "installPhase";
```
Just do
```
sourceRoot= ".";
```

and this should work. followed by fixing https://github.com/NixOS/nixpkgs/pull/66249/files#r311720855

virusdave reviewed (commented):
> This is a binary package so it needs to be patched.
> You should be able to do this by adding `autoPatchelfHook` to `nativeBuildInputs`.

I don't really understand what this is.  What needs to be patched?  Can you explain @worldofpeace  ?  Thanks!

worldofpeace commented:
I don't know what happened to this thread, but I can't read it.

I'll try responding like this:

> This doesn't work. It fails in "unpacking sources" phase, since no directory is actually created.

You need to do

```
sourceRoot= ".";
```

> > Whatever got unpacked should be in the $CWD for you to copy into $out
>
> I don't follow. Please explain?

I mean `cp ucm $out/bin`


> > This is a binary package so it needs to be patched.
> >You should be able to do this by adding autoPatchelfHook to nativeBuildInputs.
>
> I don't really understand what this is. What needs to be patched? Can you explain @worldofpeace ? Thanks!

https://nixos.wiki/wiki/Packaging/Binaries explains this very well.
Simply put the binary `ucm` won't work on NixOS without hardcoding libraries to runtime search path of the executable etc. You probably didn't notice this because this binary would work on MacOS by default.

By doing
>You should be able to do this by adding autoPatchelfHook to nativeBuildInputs.

it should take care of this for you completely, no need to patch it manually.

---

> Not sure what you'd prefer to see here. This software is called ucm, and it is (description) the Unison Codebase Manager. What do you recommend i put here?


I got this from their github
"Modern, statically-typed purely functional language"

virusdave commented:
> I don't know what happened to this thread, but I can't read it.
> 

I'm sure i messed it up somehow, but not sure how.  Sorry :(

> I'll try responding like this:
> 
> > This doesn't work. It fails in "unpacking sources" phase, since no directory is actually created.
> 
> You need to do
> 
> ```
> sourceRoot= ".";
> ```
> 
> > > Whatever got unpacked should be in the $CWD for you to copy into $out
> > 
> > 
> > I don't follow. Please explain?
> 
> I mean `cp ucm $out/bin`
> 

Gotcha.  This (and adding an `unpackPhase` to the list) together worked.

> > > This is a binary package so it needs to be patched.
> > > You should be able to do this by adding autoPatchelfHook to nativeBuildInputs.
> > 
> > 
> > I don't really understand what this is. What needs to be patched? Can you explain @worldofpeace ? Thanks!
> 
> https://nixos.wiki/wiki/Packaging/Binaries explains this very well.
> Simply put the binary `ucm` won't work on NixOS without hardcoding libraries to runtime search path of the executable etc. You probably didn't notice this because this binary would work on MacOS by default.
> 

Oh interesting.  Yeah, i hadn't noticed this.  Fixed!

> By doing
> 
> > You should be able to do this by adding autoPatchelfHook to nativeBuildInputs.
> 
> it should take care of this for you completely, no need to patch it manually.
> 
> > Not sure what you'd prefer to see here. This software is called ucm, and it is (description) the Unison Codebase Manager. What do you recommend i put here?
> 
> I got this from their github
> "Modern, statically-typed purely functional language"

Updated!

virusdave commented:
OK, i think i've made all the requested changes.  It still works on my mac system, at least. :)

worldofpeace commented on pkgs/development/compilers/unison/default.nix:
```diff
@@ -0,0 +1,39 @@
+
+  # The tarball is just the prebuilt binary, in the archive root.
+  sourceRoot = ".";
+  phases = "unpackPhase installPhase fixupPhase";
```
```suggestion
  dontBuild = true;
  dontConfigure = true;
```

virusdave replied:
Done!

worldofpeace commented:
> I'm sure i messed it up somehow, but not sure how. Sorry :(

I actually think GitHub had something to do with this.

worldofpeace commented on pkgs/development/compilers/unison/default.nix:
```diff
@@ -0,0 +1,39 @@
+    homepage = http://unisonweb.org/posts/;
+
+    license = licenses.bsd3;
+    maintainers = maintainers.virusdave;
```
```suggestion
    maintainers = [ maintainers.virusdave ];
```

fixes eval failure, need to be a list

virusdave replied:
Done!

virusdave commented:
@worldofpeace 
Is there any other changes you think i should make?

Thanks again for taking the time to review this!

worldofpeace commented:
Can you change the description to just?
"Modern, statically-typed purely functional language"

It appears you've intersected two of them, and it seems like a lot of adjectives.

And I don't think we should use whitespace between the `meta` attributes.

virusdave commented:
> Can you change the description to just?
> "Modern, statically-typed purely functional language"
> 
> It appears you've intersected two of them, and it seems like a lot of adjectives.

Done!

> 
> And I don't think we should use whitespace between the `meta` attributes.
> 

Done!

> Lastly, it fails to build on NixOS because of missing dependencies
> 
> ```
> libz.so.1 -> not found!
> libtinfo.so.5 -> not found!
> libgmp.so.10 -> not found!
> ```
> 
> so you need
> 
> * `zlib`
> * `ncurses`
> * `gpm`

Hmm, i guess this makes sense for nixos.  Shared libraries need to be declared, right?  
I've added the runtime dependencies.  

Done!

worldofpeace commented:
> Hmm, i guess this makes sense for nixos. Shared libraries need to be declared, right?
I've added the runtime dependencies.

That output is from `autoPatchelfHook` and since you've built it on MacOS without sandboxing it probably found those shared libraries from the system.

worldofpeace commented on pkgs/development/compilers/unison/default.nix:
```diff
@@ -0,0 +1,40 @@
+{ stdenv, fetchurl, autoPatchelfHook
+, zlib, ncurses, gpm
```
Sorry, forgot to mention you have to add these to
`buildInputs`.

virusdave replied:
OK, done.  Still builds on osx, except i can't have `gpm` as a `buildInput` since it's not supported on osx.

worldofpeace commented:
@GrahamcOfBorg build unison-ucm

worldofpeace commented:
It looks like I've forgotten that obviously you don't need to patch binaries like this for darwin, and I miss suggested `gpm` when I meant `gmp` which is what haskell use's for multiple precision arithmetic.

worldofpeace commented:
Thanks for contributing @virusdave :sparkles: 
The above issue I fixed in the force push.

virusdave commented:
Thanks for the review @worldofpeace and your patience with me.  This is my first "full package derivation" experience, and although it was simple (binary distro even), it was useful learning experience and i appreciate the help :)

worldofpeace commented:
> Thanks for the review @worldofpeace and your patience with me. This is my first "full package derivation" experience, and although it was simple (binary distro even), it was useful learning experience and i appreciate the help :)

I'm happy to hear this :cherry_blossom: . 
I like to think the nix manual's are helpful but there's nothing like the people behind them that write them. Always be prepared to learn something new from the people that are here :sparkles:
